### ISSUE TITLE
NodeJS v7.0.0 broken because of removal of WeakCallbackData in V8

### ISSUE THREAD

**Issue by gpean** [swig/swig#825]
Swig wrapper code makes use of WeakCallbackData which has been deprecated and is now removed in node v7.0.0. I think we have to move to WeakCallbackInfo instead, but I didn't look into it further.

**Comment by arfoll** [comment]
It seems that it's also GetHiddenValue -> GetPrivate and SetHiddenValue -> GetHiddenValue that breaks it with v8 version => 0x054500. Its thi that I'm not sure how to treat now. I was hoping it would be like:

```
#if (V8_MAJOR_VERSION-0) < 4 && (SWIG_V8_VERSION < 0x054500)
  v8::Local<v8::Value> moduleinfo = global_obj->GetHiddenValue(SWIGV8_STRING_NEW("swig_module_info_data"));
#else
  v8::Local<v8::Value> moduleinfo;
  v8::Local<v8::Private> privateKey = v8::Private::ForApi(v8::Isolate::GetCurrent(), SWIGV8_STRING_NEW("swig_module_info_data"));
  global_obj->GetPrivate(SWIGV8_CURRENT_CONTEXT(), privateKey).ToLocal(&moduleinfo);
#endif
```

**Comment by g-vidal** [comment]
Hi,
I came across the same problem compiling on my laptop, here are the messages in case it may help 
```shell
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:966:27: error: ‘WeakCallbackData’ in namespace ‘v8’ does not name a template type
   void (*dtor) (const v8::WeakCallbackData<v8::Object, SWIGV8_Proxy> &data);
                           ^~~~~~~~~~~~~~~~
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:966:43: error: expected ‘,’ or ‘...’ before ‘<’ token
   void (*dtor) (const v8::WeakCallbackData<v8::Object, SWIGV8_Proxy> &data);
                                           ^
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1014:53: error: ‘WeakCallbackData’ in namespace ‘v8’ does not name a template type
 SWIGRUNTIME void SWIGV8_Proxy_DefaultDtor(const v8::WeakCallbackData<v8::Object, SWIGV8_Proxy> &data) {
                                                     ^~~~~~~~~~~~~~~~
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1014:69: error: expected ‘,’ or ‘...’ before ‘<’ token
 SWIGRUNTIME void SWIGV8_Proxy_DefaultDtor(const v8::WeakCallbackData<v8::Object, SWIGV8_Proxy> &data) {
                                                                     ^
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx: In function ‘void SWIGV8_Proxy_DefaultDtor(int)’:
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1015:25: error: ‘data’ was not declared in this scope
   SWIGV8_Proxy *proxy = data.GetParameter();
```
Then a bit further
```shell
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1086:78: error: no matching function for call to ‘v8::Persistent<v8::Object>::SetWeak(SWIGV8_Proxy*&, void (*&)(int))’
     cdata->handle.SetWeak(cdata, ((SWIGV8_ClientData*)info->clientdata)->dtor);
```
```shell
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1088:58: error: no matching function for call to ‘v8::Persistent<v8::Object>::SetWeak(SWIGV8_Proxy*&, void (&)(int))’
     cdata->handle.SetWeak(cdata, SWIGV8_Proxy_DefaultDtor);
```
And finally

```shell
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx: In function ‘int SwigV8Packed_Check(v8::Handle<v8::Value>)’:
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1242:40: error: ‘class v8::Object’ has no member named ‘GetHiddenValue’; did you mean ‘IntegerValue’?
   v8::Handle<v8::Value> flag = objRef->GetHiddenValue(SWIGV8_STRING_NEW("__swig__packed_data__"));
                                        ^~~~~~~~~~~~~~
[ 84%] Built target blink-io-cpp
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx: At global scope:
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1292:58: error: ‘WeakCallbackData’ in namespace ‘v8’ does not name a template type
 SWIGRUNTIME void _wrap_SwigV8PackedData_delete(const v8::WeakCallbackData<v8::Object, SwigV8PackedData> &data) {
                                                          ^~~~~~~~~~~~~~~~
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1292:74: error: expected ‘,’ or ‘...’ before ‘<’ token
 SWIGRUNTIME void _wrap_SwigV8PackedData_delete(const v8::WeakCallbackData<v8::Object, SwigV8PackedData> &data) {
                                                                          ^
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx: In function ‘void _wrap_SwigV8PackedData_delete(int)’:
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1293:34: error: ‘data’ was not declared in this scope
   v8::Local<v8::Object> object = data.GetValue();
                                  ^~~~
[ 84%] Built target Spi-pot
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx: In function ‘v8::Handle<v8::Value> SWIGV8_NewPackedObj(void*, size_t, swig_type_info*)’:
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1322:8: error: ‘class v8::Object’ has no member named ‘SetHiddenValue’; did you mean ‘IntegerValue’?
   obj->SetHiddenValue(SWIGV8_STRING_NEW("__swig__packed_data__"), SWIGV8_BOOLEAN_NEW(true));
        ^~~~~~~~~~~~~~
/home/vidal/Logiciels/IoT/mraa/build/src/javascript/mraajsJAVASCRIPT_wrap.cxx:1346:61: error: no matching function for call to ‘v8::Persistent<v8::Object>::SetWeak(SwigV8PackedData*&, void (&)(int))’
   cdata->handle.SetWeak(cdata, _wrap_SwigV8PackedData_delete);
```
As @arfoll says I can confirm that the problem occurs with node-v7.1.0 but disappears with node-v6.9.1.
Is there any schedule to fix the bug and move on towards node7.

thanks again for the work!

**Comment by arfoll** [comment]
Well this makes it build - https://github.com/arfoll/swig/tree/node7. But it segfaults immediately. I don't understand how these v8 calls well enough.

**Comment by ojwb** [comment]
> Is there any schedule to fix the bug and move on towards node7.

No.

There doesn't seem to be an active maintainer for the javascript backend (or any of its sub-backends) currently, so for now fixes and enhancements are reliant on people submitting good quality patches with appropriate test coverage, so that a SWIG maintainer without javascript familiarity feels comfortable about applying them.

**Comment by g-vidal** [comment]
Hi all,
I have tried with node 7.2  the problem is still there, unfortunatenately I have no skills to propose a patch but I am a little worried because it seems that we are stuck behind version 6 whereas  node is progressing fast which may raise bigger and time-consuming side effects .
REally sorry not being able to do more than pushing up the topic.

**Comment by ojwb** [comment]
I noticed we already have #804 for this same issue, so closing this as a duplicate.

**Comment by SlickWish** [comment]
It doesn't just cause problems for Node7, but for the current version of Electron too. Electron uses 5.4 v8.

Is anyone actively working on this?

**Comment by ojwb** [comment]
This issue has been closed as a duplicate - please use #804 for discussion instead.
